### PR TITLE
[CLOSE ME] Switch to multi-asset coin selection algorithm

### PR DESCRIPTION
**Pull request by errfrom**
This PR has the wrong base branch, see https://github.com/Plutonomicon/cardano-transaction-lib/pull/1268.

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [ ] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [ ] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: src/Internal/BalanceTx/BalanceTx.purs [hunk 1]
```diff
@@ -290,215 +351,306 @@ addLovelacesToTransactionOutput txOutput = do
...
+
+    fromArrayUnsafe :: forall (a :: Type). Array a -> NonEmptyArray a
+    fromArrayUnsafe = unsafePartial fromJust <<< NEArray.fromArray
+
+    noTokens :: Value /\ BigInt -> Boolean
+    noTokens = Array.null <<< Value.valueAssets <<< fst
+
+  adaRequiredAtStart :: BalanceTxM BigInt
+  adaRequiredAtStart =
+    foldr (+) zero <$> traverse (minCoinFor <<< fst) pairsAtStart
```
Comment (klntsky) [inline src/Internal/BalanceTx/BalanceTx.purs:554]:
```suggestion
    sum <$> traverse (minCoinFor <<< fst) pairsAtStart
```

**Review by klntsky — CHANGES_REQUESTED** [review]
We definitely need more tests. Maybe we can try running some of the test contracts with the new strategy.

File: src/Internal/BalanceTx/BalanceTx.purs [hunk 2]
```diff
@@ -197,70 +219,109 @@ balanceTxWithConstraints unbalancedTx constraintsBuilder =
...
+      if requiredValue == mempty then balanceChangeAndMinFee else prebalanceTx
+
+  -- | Selects a combination of unspent transaction outputs from the wallet's
+  -- | utxo set so that the total input value is sufficient to cover all 
+  -- | transaction outputs, including generated change and min fee.
+  prebalanceTx :: BalancerState -> BalanceTxM FinalizedTransaction
+  prebalanceTx state@{ unbalancedTx, changeOutputs, leftoverUtxos } = do
+    logBalancerState "Pre-balancing (Stage 1)" utxos state
+
+    performCoinSelection >>= \selectionState ->
```
Comment (jy14898) [inline src/Internal/BalanceTx/BalanceTx.purs:272]: Might as well just use do notation here
```suggestion
    selectionState <- performCoinSelection
```

**Comment by errfrom** [comment]
@klntsky The new selection strategy is now used by default, so all our test contracts are executed using it.

File: src/Internal/BalanceTx/CoinSelection.purs [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+      Just { head: p, tail: ps' } ->
+        p s >>= case _ of
+          Nothing -> go s ps' qs
+          Just s' -> go (demote s') ps' (Array.snoc qs p)
+
+--------------------------------------------------------------------------------
+-- SelectionPriority
+--------------------------------------------------------------------------------
+
+data SelectionPriority = SelectionPriorityCover | SelectionPriorityImprove
```
Comment (klntsky) [inline src/Internal/BalanceTx/CoinSelection.purs:381]: Please describe how these alternatives affect the algorithm

**Review by klntsky — COMMENTED** [review]
Overall very good, I will return to it later for a more in-depth review

**Comment by klntsky** [comment]
Do not merge, the target is wrong.

**Comment by uhbif19** [comment]
@klntsky @errfrom Closing, cuz there is newer PR.

**Pull request closed without merging**